File: backend/app/api/v1/routers/records.py
import json
import unicodedata
import uuid
from urllib.parse import quote

from fastapi import APIRouter, Depends, Form, HTTPException, Query, UploadFile, status
from fastapi.responses import StreamingResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user, get_org_context
from app.dynamic_fields.validation import FieldValidationError
from app.models.membership import Membership
from app.models.record import RecordEvent
from app.models.user import User
from app.schemas.record import (
    RecordCreate,
    RecordEventCreate,
    RecordEventOut,
    RecordListOut,
    RecordOut,
    RecordUpdate,
    RecordUpdateOut,
)
from app.schemas.search import ImportCommitResult, ImportMappingSuggestion
from app.services.export_service import export_records_csv
from app.services.import_service import ImportParseError, build_rows, parse_spreadsheet, suggest_mapping
from app.services.model_definition_service import ModelDefinitionService
from app.services.organization_service import PermissionDeniedError
from app.services.record_service import RecordService

router = APIRouter(tags=["records"])


def _validation_error_response(exc: FieldValidationError) -> HTTPException:
    return HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, {"errors": exc.errors})


def _parse_filters(raw: str | None) -> dict[str, str] | None:
    if not raw:
        return None
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Le paramètre filters doit être un objet JSON.") from exc
    if not isinstance(parsed, dict) or not all(isinstance(v, str) for v in parsed.values()):
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "filters doit être un objet {champ: valeur texte}.")
    return parsed


@router.get("/organizations/{organization_id}/model-definitions/{model_id}/records", response_model=RecordListOut)
async def list_records(
    model_id: uuid.UUID,
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
    status_filter: str | None = Query(default=None, alias="status"),
    include_archived: bool = Query(default=False),
    filters: str | None = Query(default=None, description='JSON {"champ": "valeur"} — cahier des charges §9'),
    sort_key: str | None = Query(default=None),
    sort_direction: str = Query(default="desc", pattern="^(asc|desc)$"),
    limit: int = Query(default=50, le=200),
    offset: int = Query(default=0, ge=0),
) -> RecordListOut:
    model_service = ModelDefinitionService(db)
    model = await model_service.get(membership.organization_id, model_id)
    if model is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Modèle introuvable.")

    record_service = RecordService(db)
    items, total = await record_service.list_for_model(
        membership.organization_id,
        model_id,
        include_archived=include_archived,
        status=status_filter,
        field_filters=_parse_filters(filters),
        sort_key=sort_key,
        sort_direction=sort_direction,
        limit=limit,
        offset=offset,
    )
    return RecordListOut(
        items=[RecordOut.model_validate(r) for r in items], total=total, limit=limit, offset=offset
    )


@router.post(
    "/organizations/{organization_id}/model-definitions/{model_id}/records",
    response_model=RecordOut,
    status_code=status.HTTP_201_CREATED,
)
async def create_record(
    model_id: uuid.UUID,
    payload: RecordCreate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> RecordOut:
    model_service = ModelDefinitionService(db)
    model = await model_service.get(membership.organization_id, model_id)
    if model is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Modèle introuvable.")

    record_service = RecordService(db)
    try:
        record = await record_service.create(
            organization_id=membership.organization_id,
            actor=user,
            actor_membership=membership,
            model=model,
            data=payload.data,
            status=payload.status,
            site=payload.site,
            assigned_person_record_id=payload.assigned_person_record_id,
            record_id=payload.id,
        )
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc
    except FieldValidationError as exc:
        raise _validation_error_response(exc) from exc
    return RecordOut.model_validate(record)


@router.get("/organizations/{organization_id}/records/{record_id}", response_model=RecordOut)
async def get_record(
    record_id: uuid.UUID, membership: Membership = Depends(get_org_context), db: AsyncSession = Depends(get_db)
) -> RecordOut:
    record = await RecordService(db).get(membership.organization_id, record_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Fiche introuvable.")
    return RecordOut.model_validate(record)


@router.patch("/organizations/{organization_id}/records/{record_id}", response_model=RecordUpdateOut)
async def update_record(
    record_id: uuid.UUID,
    payload: RecordUpdate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> RecordUpdateOut:
    record_service = RecordService(db)
    record = await record_service.get(membership.organization_id, record_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Fiche introuvable.")

    model = await ModelDefinitionService(db).get(membership.organization_id, record.model_definition_id)
    if model is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Modèle introuvable.")

    try:
        record = await record_service.update(
            organization_id=membership.organization_id,
            actor=user,
            actor_membership=membership,
            model=model,
            record=record,
            data=payload.data,
            status=payload.status,
            site=payload.site,
            assigned_person_record_id=payload.assigned_person_record_id,
            client_operation_id=payload.client_operation_id,
            field_written_at=payload.field_written_at,
        )
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc
    except FieldValidationError as exc:
        raise _validation_error_response(exc) from exc
    return RecordUpdateOut(
        **RecordOut.model_validate(record).model_dump(),
        conflicted_field_keys=getattr(record, "conflicted_field_keys", []),
    )


@router.post("/organizations/{organization_id}/records/{record_id}/archive", response_model=RecordOut)
async def archive_record(
    record_id: uuid.UUID,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> RecordOut:
    record_service = RecordService(db)
    record = await record_service.get(membership.organization_id, record_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Fiche introuvable.")
    try:
        record = await record_service.archive(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, record=record
        )
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc
    return RecordOut.model_validate(record)


@router.get("/organizations/{organization_id}/records/{record_id}/events", response_model=list[RecordEventOut])
async def list_record_events(
    record_id: uuid.UUID, membership: Membership = Depends(get_org_context), db: AsyncSession = Depends(get_db)
) -> list[RecordEventOut]:
    stmt = (
        select(RecordEvent)
        .where(RecordEvent.record_id == record_id, RecordEvent.organization_id == membership.organization_id)
        .order_by(RecordEvent.occurred_at.desc())
    )
    events = (await db.execute(stmt)).scalars().all()
    return [RecordEventOut.model_validate(e) for e in events]


@router.post(
    "/organizations/{organization_id}/records/{record_id}/events",
    response_model=RecordEventOut,
    status_code=status.HTTP_201_CREATED,
)
async def add_record_event(
    record_id: uuid.UUID,
    payload: RecordEventCreate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> RecordEventOut:
    record_service = RecordService(db)
    record = await record_service.get(membership.organization_id, record_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Fiche introuvable.")
    try:
        event = await record_service.add_event(
            organization_id=membership.organization_id,
            actor=user,
            actor_membership=membership,
            record=record,
            payload=payload,
        )
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc
    return RecordEventOut.model_validate(event)


# --- export / import (cahier des charges §9) ------------------------------------------

# Borne haute pour un export en une seule réponse ; au-delà, un export en tâche de
# fond serait nécessaire (§14.3) — non construit dans ce lot, volume jugé suffisant
# pour les organisations visées (voir PRODUCT.md §4, hypothèse Q6).
_EXPORT_ROW_LIMIT = 10_000


#: Caractères qu'un nom de fichier ne peut pas porter dans un en-tête HTTP : le
#: guillemet fermerait la valeur, l'antislash l'échapperait. Les caractères de
#: contrôle sont écartés à part, par `str.isprintable()`.
_UNSAFE_IN_HEADER = '"\\'


def _content_disposition(filename: str) -> str:
    """En-tête de pièce jointe conforme au RFC 6266.

    Le nom du modèle est libre — il vient du client. L'écrire directement dans
    l'en-tête (`filename="{nom}.csv"`) marchait tant que ce nom tenait dans le
    latin-1, seul jeu de caractères que les en-têtes HTTP savent transporter en
    clair : « Véhicules » passe. Un modèle nommé en cyrillique, en arabe, ou avec
    un simple « € », faisait lever un `UnicodeEncodeError` au moment d'écrire la
    réponse — un export en erreur 500, sans rapport visible avec la cause.

    On envoie donc les deux formes, comme le prescrit le RFC : un `filename`
    ASCII de repli pour les clients anciens, et un `filename*` encodé en UTF-8
    que tous les navigateurs actuels préfèrent.
    """
    stem, dot, extension = filename.rpartition(".")
    if not dot:  # aucun point : tout est le radical
        stem, extension = filename, ""

    ascii_stem = unicodedata.normalize("NFKD", stem).encode("ascii", "ignore").decode()
    # Un saut de ligne est de l'ASCII : il survit à `encode("ascii")` et
    # permettrait d'injecter un en-tête HTTP entier dans la réponse. Ces
    # caractères viennent d'un nom saisi par l'utilisateur — on les retire.
    ascii_stem = "".join(c for c in ascii_stem if c.isprintable() and c not in _UNSAFE_IN_HEADER).strip()

    # Le repli se calcule sur le RADICAL et non sur le nom entier : « 🚚.csv »
    # laissait sinon « .csv », c'est-à-dire un fichier caché sans nom, que le
    # système d'exploitation refuse ou masque.
    ascii_extension = "".join(c for c in extension if c.isalnum())
    safe_name = f"{ascii_stem or 'export'}.{ascii_extension or 'csv'}"

    quoted = quote(filename, safe="")
    return f'attachment; filename="{safe_name}"; filename*=UTF-8\'\'{quoted}'


@router.get("/organizations/{organization_id}/model-definitions/{model_id}/records/export.csv")
async def export_records(
    model_id: uuid.UUID,
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
    filters: str | None = Query(default=None),
    columns: str | None = Query(default=None, description="Clés de champs séparées par des virgules"),
) -> StreamingResponse:
    model_service = ModelDefinitionService(db)
    model = await model_service.get(membership.organization_id, model_id)
    if model is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Modèle introuvable.")

    record_service = RecordService(db)
    items, _ = await record_service.list_for_model(
        membership.organization_id, model_id, field_filters=_parse_filters(filters), limit=_EXPORT_ROW_LIMIT, offset=0
    )
    column_keys = columns.split(",") if columns else None
    csv_content = export_records_csv(model, items, columns=column_keys)
    return StreamingResponse(
        iter([csv_content]),
        media_type="text/csv",
        headers={"Content-Disposition": _content_disposition(f"{model.name_plural}.csv")},
    )


@router.post(
    "/organizations/{organization_id}/model-definitions/{model_id}/records/import/preview",
    response_model=ImportMappingSuggestion,
)
async def preview_import(
    model_id: uuid.UUID,
    file: UploadFile,
    mapping: str | None = Form(default=None),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> ImportMappingSuggestion:
    """`mapping` (optionnel) : la correspondance déjà corrigée à l'écran. Sans lui,
    l'aperçu est calculé sur la correspondance suggérée automatiquement.

    L'accepter est ce qui rend les compteurs honnêtes : dès que l'utilisateur
    corrige une colonne, « N lignes valides » doit décrire *sa* correspondance, pas
    celle devinée au départ.
    """
    model_service = ModelDefinitionService(db)
    model = await model_service.get(membership.organization_id, model_id)
    if model is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Modèle introuvable.")

    try:
        sheet = parse_spreadsheet(await file.read(), file.filename, file.content_type)
    except ImportParseError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, str(exc)) from exc

    suggested = suggest_mapping(sheet.headers, model.field_definitions)
    if mapping is None:
        applied = {h: k for h, k in suggested.items() if k}
    else:
        try:
            applied = json.loads(mapping)
        except json.JSONDecodeError as exc:
            raise HTTPException(status.HTTP_400_BAD_REQUEST, "mapping doit être un objet JSON.") from exc

    rows = build_rows(sheet.rows, applied, model.field_definitions)

    return ImportMappingSuggestion(
        headers=sheet.headers,
        suggested_mapping=suggested,
        preview_rows=sheet.rows[:10],
        total_rows=len(rows),
        valid_row_count=sum(1 for r in rows if r.is_valid),
        invalid_row_count=sum(1 for r in rows if not r.is_valid),
        sample_errors=[{"row": r.index, "errors": r.errors} for r in rows if not r.is_valid][:20],
        source_format=sheet.source_format,
        sheet_name=sheet.sheet_name,
        ignored_sheet_names=sheet.ignored_sheet_names,
    )


@router.post(
    "/organizations/{organization_id}/model-definitions/{model_id}/records/import/commit",
    response_model=ImportCommitResult,
)
async def commit_import(
    model_id: uuid.UUID,
    file: UploadFile,
    mapping: str = Form(...),
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> ImportCommitResult:
    """`mapping` : JSON {"en_tete_colonne": "cle_de_champ"} — celui validé par l'utilisateur
    à l'écran d'aperçu (§9 : « correspondance des colonnes et aperçu avant validation »).
    """
    model_service = ModelDefinitionService(db)
    model = await model_service.get(membership.organization_id, model_id)
    if model is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Modèle introuvable.")
    try:
        mapping_dict = json.loads(mapping)
    except json.JSONDecodeError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "mapping doit être un objet JSON.") from exc

    try:
        sheet = parse_spreadsheet(await file.read(), file.filename, file.content_type)
    except ImportParseError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, str(exc)) from exc
    rows = build_rows(sheet.rows, mapping_dict, model.field_definitions)

    record_service = RecordService(db)
    created = 0
    errors: list[dict] = []
    for row in rows:
        if not row.is_valid:
            errors.append({"row": row.index, "errors": row.errors})
            continue
        try:
            await record_service.create(
                organization_id=membership.organization_id,
                actor=user,
                actor_membership=membership,
                model=model,
                data=row.data,
                status=None,
                site=None,
                assigned_person_record_id=None,
            )
            created += 1
        except (PermissionDeniedError, FieldValidationError) as exc:
            message = exc.errors if isinstance(exc, FieldValidationError) else {"_": str(exc)}
            errors.append({"row": row.index, "errors": message})

    return ImportCommitResult(created=created, failed=len(errors), errors=errors)

File: frontend/src/lib/api/records.ts
import { apiRequest, apiRequestRaw } from "./http";
import type {
  ImportCommitResult,
  ImportMappingSuggestion,
  RecordCreate,
  RecordEventCreate,
  RecordEventOut,
  RecordListOut,
  RecordOut,
  RecordUpdate,
  RecordUpdateOut,
} from "./types";

/**
 * Client des fiches (cahier des charges §6). Même pattern que `organizations.ts` :
 * `accessToken` en premier paramètre, `apiRequest` lève `ApiError` sur tout échec.
 *
 * Pagination : `listRecords` est **toujours** appelée avec `limit`/`offset` — jamais
 * de "charger tout puis paginer côté client". C'est ce qui permet à la vue liste de
 * rester rapide avec 10 000 fiches (cahier des charges §14.3). `RecordListOut.total`
 * porte le nombre réel de fiches côté serveur, utilisé pour calculer le nombre de pages.
 *
 * Tri et filtres par champ : le service backend (`RecordService.list_for_model`) sait
 * trier par `sort_key`/`sort_direction` et filtrer par `field_filters`, mais la route
 * HTTP (`GET .../records`) n'expose aujourd'hui que `status`, `include_archived`,
 * `limit`, `offset` — pas de tri ni de filtre par champ personnalisé côté API. Ne pas
 * ajouter ces paramètres ici tant que la route ne les accepte pas réellement : ça
 * afficherait un tri qui n'aurait aucun effet, un faux contrôle plutôt qu'une absence
 * honnête (recherche/filtres avancés = lot 3, voir PRODUCT.md §10.2).
 */

export interface ListRecordsParams {
  status?: string;
  includeArchived?: boolean;
  limit: number;
  offset: number;
}

export function listRecords(
  accessToken: string,
  organizationId: string,
  modelId: string,
  params: ListRecordsParams,
): Promise<RecordListOut> {
  const search = new URLSearchParams();
  if (params.status) search.set("status", params.status);
  if (params.includeArchived) search.set("include_archived", "true");
  search.set("limit", String(params.limit));
  search.set("offset", String(params.offset));
  return apiRequest<RecordListOut>(
    `/organizations/${organizationId}/model-definitions/${modelId}/records?${search.toString()}`,
    { accessToken },
  );
}

export function getRecord(
  accessToken: string,
  organizationId: string,
  recordId: string,
): Promise<RecordOut> {
  return apiRequest<RecordOut>(`/organizations/${organizationId}/records/${recordId}`, {
    accessToken,
  });
}

export function createRecord(
  accessToken: string,
  organizationId: string,
  modelId: string,
  payload: RecordCreate,
): Promise<RecordOut> {
  return apiRequest<RecordOut>(
    `/organizations/${organizationId}/model-definitions/${modelId}/records`,
    { accessToken, method: "POST", body: JSON.stringify(payload) },
  );
}

/** Renvoie `RecordUpdateOut` (pas `RecordOut`) : porte `conflicted_field_keys`,
 * les clés rejetées par la fusion champ par champ de cet appel précis (§11.3). */
export function updateRecord(
  accessToken: string,
  organizationId: string,
  recordId: string,
  payload: RecordUpdate,
): Promise<RecordUpdateOut> {
  return apiRequest<RecordUpdateOut>(`/organizations/${organizationId}/records/${recordId}`, {
    accessToken,
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function archiveRecord(
  accessToken: string,
  organizationId: string,
  recordId: string,
): Promise<RecordOut> {
  return apiRequest<RecordOut>(
    `/organizations/${organizationId}/records/${recordId}/archive`,
    { accessToken, method: "POST" },
  );
}

/**
 * Import initial (cahier des charges §9) — deux temps, deux appels, le **même
 * fichier** envoyé aux deux : le backend ne garde rien entre l'aperçu et la
 * validation, il relit le fichier à chaque fois.
 *
 * Le corps est un `FormData` : `apiRequest` omet alors son `Content-Type` JSON
 * pour laisser le navigateur poser le `multipart/form-data; boundary=...`
 * (voir `http.ts`). `mapping` part en champ de formulaire *encodé en JSON*, pas
 * en query string — même contrainte FastAPI que `field_key` sur l'envoi de
 * document : une route qui mêle `UploadFile` et `Form(...)` lit tout dans le corps.
 */
export function previewImport(
  accessToken: string,
  organizationId: string,
  modelId: string,
  file: File,
  mapping?: Record<string, string>,
): Promise<ImportMappingSuggestion> {
  const form = new FormData();
  form.set("file", file);
  // Sans `mapping`, l'aperçu est calculé sur la correspondance suggérée. Dès que
  // l'utilisateur en corrige une, on la renvoie : les compteurs doivent décrire
  // la correspondance réellement retenue, pas celle devinée au départ.
  if (mapping) form.set("mapping", JSON.stringify(mapping));
  return apiRequest<ImportMappingSuggestion>(
    `/organizations/${organizationId}/model-definitions/${modelId}/records/import/preview`,
    { accessToken, method: "POST", body: form },
  );
}

/** `mapping` : {en-tête de colonne: clé de champ}. N'y mettre que les colonnes
 * réellement importées — une colonne absente est ignorée par le backend. */
export function commitImport(
  accessToken: string,
  organizationId: string,
  modelId: string,
  file: File,
  mapping: Record<string, string>,
): Promise<ImportCommitResult> {
  const form = new FormData();
  form.set("file", file);
  form.set("mapping", JSON.stringify(mapping));
  return apiRequest<ImportCommitResult>(
    `/organizations/${organizationId}/model-definitions/${modelId}/records/import/commit`,
    { accessToken, method: "POST", body: form },
  );
}

export function listRecordEvents(
  accessToken: string,
  organizationId: string,
  recordId: string,
): Promise<RecordEventOut[]> {
  return apiRequest<RecordEventOut[]>(
    `/organizations/${organizationId}/records/${recordId}/events`,
    { accessToken },
  );
}

export function addRecordEvent(
  accessToken: string,
  organizationId: string,
  recordId: string,
  payload: RecordEventCreate,
): Promise<RecordEventOut> {
  return apiRequest<RecordEventOut>(
    `/organizations/${organizationId}/records/${recordId}/events`,
    { accessToken, method: "POST", body: JSON.stringify(payload) },
  );
}

/** Plafond du serveur (`_EXPORT_ROW_LIMIT`). Répliqué ici pour que l'écran
 *  puisse *prévenir* avant l'export plutôt que de laisser l'utilisateur
 *  découvrir un fichier tronqué. Doit rester aligné sur le backend. */
export const EXPORT_ROW_LIMIT = 10_000;

export interface ExportedFile {
  blob: Blob;
  filename: string;
}

/**
 * Export CSV des fiches d'un modèle (cahier des charges §10 — « consulter et
 * exporter » est un droit du rôle Lecteur lui-même).
 *
 * La route existait côté backend depuis le lot 3, **sans aucun appelant côté
 * frontend** : la fonction était donc inatteignable depuis l'application.
 *
 * Passe par `apiRequestRaw` : l'authentification voyage dans un en-tête, ce
 * qu'un simple `<a href>` ne sait pas faire. Le fichier est donc récupéré en
 * mémoire puis remis au navigateur — acceptable, l'export est plafonné à
 * `EXPORT_ROW_LIMIT` lignes.
 */
export async function exportRecordsCsv(
  accessToken: string,
  organizationId: string,
  modelId: string,
  options: { filters?: string; columns?: string[] } = {},
): Promise<ExportedFile> {
  const search = new URLSearchParams();
  if (options.filters) search.set("filters", options.filters);
  if (options.columns?.length) search.set("columns", options.columns.join(","));
  const query = search.toString();

  const response = await apiRequestRaw(
    `/organizations/${organizationId}/model-definitions/${modelId}/records/export.csv${query ? `?${query}` : ""}`,
    { accessToken, headers: { "X-Organization-Id": organizationId } },
  );

  return { blob: await response.blob(), filename: filenameFromDisposition(response) };
}

/**
 * Nom de fichier proposé par le serveur (`Content-Disposition`).
 *
 * Le repli n'est pas décoratif : le navigateur n'expose cet en-tête au script
 * que si le serveur l'autorise explicitement (`Access-Control-Expose-Headers`).
 * Sans repli, un déploiement mal configuré ferait télécharger un fichier nommé
 * « undefined ».
 */
function filenameFromDisposition(response: Response): string {
  const disposition = response.headers.get("Content-Disposition") ?? "";

  // `filename*` D'ABORD, et pas « la première forme rencontrée ».
  // Le serveur envoie les deux, dans l'ordre imposé par le RFC 6266 : le repli
  // ASCII (`filename="Vehicules.csv"`) puis la forme UTF-8
  // (`filename*=UTF-8''V%C3%A9hicules.csv`). Une expression rationnelle qui
  // accepte les deux indifféremment retient donc toujours le repli — et
  // l'accent est perdu alors même que le serveur l'avait transmis.
  const encoded = /filename\*=\s*UTF-8''([^;]+)/i.exec(disposition);
  if (encoded?.[1]) {
    try {
      return decodeURIComponent(encoded[1].trim());
    } catch {
      return encoded[1].trim();
    }
  }

  const plain = /filename=\s*"?([^";]+)"?/i.exec(disposition);
  if (plain?.[1]) return plain[1].trim();

  // Ni l'une ni l'autre : l'en-tête n'est pas exposé par CORS
  // (`Access-Control-Expose-Headers`), ou le serveur ne l'a pas envoyé.
  return "export.csv";
}
